### PR TITLE
Add/echo10 granules support sb

### PR DESCRIPTION
**Pull request by svbagwell**
Checks added:

- Range Date Time Logic Check 1
- Range Date Time Logic Check 2
- Datetime ISO Format Check
- URL Health and Status Check
- Delete Time Check
- Data type control Check
- Characteristic Name Check
- Characteristic Value Check
- MIME type Check
- Online Access URL Presence Check
- Default Date Check
- Granule Instrument Short Name Check
- Granule Platform Short Name Check
- Granule Sensor Short Name Check
- Granule Platform Presence Check
- Granule Instrument Presence Check

### CODE REVIEW

**Review by esr0004 — COMMENTED** [review]
For the “Data type control Check” (data_type_control_check), recommend removing the echo-g field "Granule/AdditionalAttributes/AdditionalAttribute/DataType” in the rule_mapping.json file. It appears to only apply at the ECHO-10 collection level.

For the “Characteristic Name Check” and “Characteristic Value Check” (characteristic_name_check, characteristic_value_check), the “Characteristics” element appears to only apply for Instruments and Sensors for ECHO-10 granules. Recommend removing the following echo-g fields in the rule_mapping.json file:

(Characteristic Name Check)
"Granule/Platforms/Platform/Characteristics/Characteristic/Name"
"Granule/Platforms/Platform/Characteristics/Characteristic"
 
(Characteristic Value Check)
"Granule/Platforms/Platform/Characteristics/Characteristic/Value",
"Granule/Platforms/Platform/Characteristics/Characteristic"

(Not completely sure on this one. Feel free to ignore this recommendation if it is incorrect.)

For the “Default Date Check” (default_date_check), recommend adding the following fields under echo-g in the rule_mapping.json file:
"Granule/Temporal/RangeDateTime/BeginningDateTime"
"Granule/Temporal/RangeDateTime/EndingDateTime"
"Granule/Temporal/SingleDateTime"
"Granule/DeleteTime"
"Granule/DataGranule/ProductionDateTime"
"Granule/OrbitCalculatedSpatialDomains/OrbitCalculatedSpatialDomain/EquatorCrossingDateTime"

File: pyQuARC/schemas/rule_mapping.json [hunk 1]
```diff
@@ -2353,6 +2485,54 @@
...
+                        "Granule/Collection/ShortName",
+                        "Granule/Collection/VersionId",
+                        "Granule/Collection/DataSetId"
+                    ]
+                }
+            ],
+            "umm-g": [
+                {
+                    "fields": [
+                        "Platform/Instrument/ShortName",
```
Comment (spa0002) [inline pyQuARC/schemas/rule_mapping.json:2506]: Platforms/Instruments/ShortName

File: pyQuARC/schemas/rule_mapping.json [hunk 2]
```diff
@@ -2353,6 +2485,54 @@
...
+                {
+                    "fields": [
+                        "Granule/Platforms/Platform/Instruments/Instrument/ShortName"
+                    ]
+                }
+            ],
+            "umm-g": [
+                {
+                    "fields": [
+                        "Platform/Instrument/ShortName"
```
Comment (spa0002) [inline pyQuARC/schemas/rule_mapping.json:2482]: Platforms/Instruments/ShortName

File: pyQuARC/schemas/rule_mapping.json [hunk 3]
```diff
@@ -4465,6 +4671,108 @@
...
+                        "Granule/Collection/ShortName",
+                        "Granule/Collection/VersionId",
+                        "Granule/Collection/DataSetId"
+                    ]
+                }
+            ],
+            "umm-g": [
+                {
+                    "fields": [
+                        "Platform/Instrument/ComposedOf/ShortName",
```
Comment (spa0002) [inline pyQuARC/schemas/rule_mapping.json:4749]: Platforms/Instruments/ComposedOf/ShortName

**Review by spa0002 — COMMENTED** [review]
Some fields need to be changed for UMM-G

**Review by jenny-m-wood — APPROVED** [review]

**Review by esr0004 — APPROVED** [review]

File: pyQuARC/schemas/rule_mapping.json (hunk 1 above)
Comment (svbagwell) [reply pyQuARC/schemas/rule_mapping.json]: addressed via [1fbf3de](https://github.com/NASA-IMPACT/pyQuARC/pull/196/commits/1fbf3de6cdbde3c7a3069f730af63deed646a46b)

File: pyQuARC/schemas/rule_mapping.json (hunk 2 above)
Comment (svbagwell) [reply pyQuARC/schemas/rule_mapping.json]: addressed via [1fbf3de](https://github.com/NASA-IMPACT/pyQuARC/pull/196/commits/1fbf3de6cdbde3c7a3069f730af63deed646a46b)

File: pyQuARC/schemas/rule_mapping.json (hunk 3 above)
Comment (svbagwell) [reply pyQuARC/schemas/rule_mapping.json]: addressed via [1fbf3de](https://github.com/NASA-IMPACT/pyQuARC/pull/196/commits/1fbf3de6cdbde3c7a3069f730af63deed646a46b)

**Review by spa0002 — APPROVED** [review]

File: pyQuARC/code/custom_validator.py [hunk 4]
```diff
@@ -98,6 +100,25 @@ def presence_check(*field_values):
...
+    @staticmethod
+    def granule_sensor_presence_check(sensor_values, collection_shortname=None, version=None, dataset_id=None):
+        """
+        Checks if sensor is provided at the granule level if provided at
+        collection level
+        """
+        if dataset_id == None:
+            collection  = requests.get(f'{CMR_URL}/search/collections.umm_json?short_name={collection_shortname}&version={version}').json()
+        else: 
+            collection  = requests.get(f'{CMR_URL}/search/collections.umm_json?DatasetId={dataset_id}').json()
```
Comment (xhagrg) [inline pyQuARC/code/custom_validator.py:112]: Can we refactor this? Form the url first, using if else and then get the response from the url?

File: pyQuARC/code/custom_validator.py [hunk 5]
```diff
@@ -98,6 +100,25 @@ def presence_check(*field_values):
...
+
+        instruments = collection['items'][0]['umm']['Platforms'][0]['Instruments'][0]
```
Comment (xhagrg) [inline pyQuARC/code/custom_validator.py:114]: What if there are no elements in platforms or instruments?

File: pyQuARC/code/string_validator.py [hunk 6]
```diff
@@ -158,6 +162,80 @@ def instrument_long_name_gcmd_check(value):
             "value": value,
         }
 
+    @staticmethod
+    @if_arg
+    def validate_granule_instrument_against_collection(instrument_shortname, collection_shortname=None, version=None, dataset_id=None):
```
Comment (xhagrg) [inline pyQuARC/code/string_validator.py:167]: This method could use some refactoring.

File: pyQuARC/code/string_validator.py [hunk 7]
```diff
@@ -158,6 +162,80 @@ def instrument_long_name_gcmd_check(value):
...
+        
+        if collection['feed']['entry']:
+            return {
+                "valid": True,
+                "value": sensor_shortname
+            }
+        return {
+            "valid": False,
+            "value": sensor_shortname
+        }
```
Comment (xhagrg) [inline pyQuARC/code/string_validator.py:237]: We can reduce the amount of if conditions used.

File: pyQuARC/code/string_validator.py [hunk 8]
```diff
@@ -199,6 +277,43 @@ def platform_short_long_name_consistency_check(*args):
             "value": (args[0], args[1]),
         }
 
+    @staticmethod
+    @if_arg
+    def validate_granule_platform_against_collection(platform_shortname, collection_shortname=None, version=None, dataset_id=None):
```
Comment (xhagrg) [inline pyQuARC/code/string_validator.py:282]: This method looks very similar to the previous method.

**Pull request merged**